python_modules/dagster/dagster/_core/definitions/asset_reconciliation_sensor.py

```diff
@@ -0,0 +1,348 @@
+    the materializations prior to time 3. If asset A materialized again at time 4, we would know that this materialization
+    has not been incorporated into the child asset yet.
+    """
+
+    def sensor_fn(context):
+        asset_defs_by_key = context._repository_def._assets_defs_by_key
+        source_asset_defs_by_key = context._repository_def.source_assets_by_key
+        upstream: Mapping[AssetKey, Set[AssetKey]] = _get_upstream_mapping(
+            selection=selection,
+            assets=asset_defs_by_key.values(),
+            source_assets=source_asset_defs_by_key.values(),
+        )
+
+        cursor_dict: Dict[str, float] = json.loads(context.cursor) if context.cursor else {}
+        should_materialize: List[AssetKey] = []
+        cursor_update_dict: Dict[str, float] = {}
+
+        # sort the assets topologically so that we process them in order
+        toposort_assets = list(toposort.toposort(upstream))
+        # unpack the list of sets into a list and only keep the ones we are monitoring
+        toposort_assets = [
+            asset for layer in toposort_assets for asset in layer if asset in upstream.keys()
+        ]
+
+        # determine which assets should materialize based on the materialization status of their
+        # parents
+        for a in toposort_assets:
+            a_cursor = cursor_dict.get(str(a), 0.0)
+            cursor_update_dict[str(a)] = a_cursor
+            parent_update_records = _get_parent_updates(
+                context,
+                current_asset=a,
+                parent_assets=upstream[a],
+                cursor_timestamp=a_cursor,
+                will_materialize_list=should_materialize,
+                wait_for_in_progress_runs=wait_for_in_progress_runs,
+            )
+
+            condition = all if and_condition else any
+            if condition(
+                [
+                    materialization_status
+                    for materialization_status, _ in parent_update_records.values()
+                ]
+            ):
+                should_materialize.append(a)
+                cursor_update_dict[str(a)] = max(
+                    [cursor_val for _, cursor_val in parent_update_records.values()] + [a_cursor]
+                )
+
+        if len(should_materialize) > 0:
+            context.update_cursor(json.dumps(cursor_update_dict))
+            context.cursor_has_been_updated = True
+            return RunRequest(run_key=f"{context.cursor}", asset_selection=should_materialize)
+
+    return MultiAssetSensorDefinition(
+        asset_keys=[],
+        asset_materialization_fn=sensor_fn,
+        name=name,
+        job_name="__ASSET_JOB",
+        minimum_interval_seconds=minimum_interval_seconds,
+        description=description,
+        default_status=default_status,
+    )
+
+
+@experimental
+def build_asset_reconciliation_sensor(
+    selection: AssetSelection,
+    name: str,
+    and_condition: bool = True,  # TODO better name for this parameter
```

Comment:
With `wait_for_in_progress_runs` defaulting to True, I think we can default this to false.

Figuring out the naming for this one is giving me a headache.

Maybe `wait_for_all_upstream`?